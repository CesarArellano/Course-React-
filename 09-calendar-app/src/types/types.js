
export const types = {
  uiOpenModal: '[UI] Open modal',
  uiCloseModal: '[UI] Close modal',

  eventSetActive: '[EVENT] Set active',
  eventAddNew: '[EVENT] Add new',
  eventClearActiveEvent: '[EVENT] Clear active event',
  eventUpdate: '[EVENT] Event updated'
};  