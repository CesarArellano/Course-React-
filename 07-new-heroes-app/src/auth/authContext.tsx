import { createContext } from "react";

export interface AuthState {
  user: {
    name: string;
    logged: boolean;
  }
  dispatch: React.Dispatch<any>
}

export const AuthContext = createContext({} as AuthState );