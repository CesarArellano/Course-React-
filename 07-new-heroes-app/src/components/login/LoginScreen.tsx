import React from 'react'
import { useNavigate } from 'react-router-dom'

export const LoginScreen = () => {

  const navigate = useNavigate();

  const handleLogin = () => {
    // navigate('/marvel') New Stack
    navigate('/marvel', {
      replace: true,
    })
  }

  return (
    <div className='container mt-3'>
      <h1>Login Screen</h1>
      <hr />
      <button 
        className="btn btn-primary"
        onClick={ handleLogin }
      >
        Login
      </button>
    </div>
  )
}
