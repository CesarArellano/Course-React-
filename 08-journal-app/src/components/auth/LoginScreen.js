import React from 'react'
import { Link } from 'react-router-dom'

export const LoginScreen = () => {
  return (
    <>
      <div className="center-box">
        <img 
          src="https://image.flaticon.com/icons/png/512/3235/3235042.png"
          alt="Journal App Icon" 
          title="App Icon"
          className="img-icon"
        />
      </div>
      <h3 className="auth__title">Login</h3>
      <form>
        <input
          type="text"
          placeholder="Email"
          name="email"
          autoComplete="off"
          className="auth__input"
        />

        <input 
          type="password"
          placeholder="Password"
          name="password"
          className="auth__input"
        />

        <div className="center-box">
          <button
            type="submit"
            className="btn btn-primary mb-5"
          >
            Login
          </button>
        </div>
        <hr />
        <div className="auth__social-networks">
          <p>Login with social networks</p>
          <div className="google-btn">
              <div className="google-icon-wrapper">
                  <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
              </div>
              <p className="btn-text">
                  <b>Sign in with google</b>
              </p>
          </div>
        </div>
          <Link 
            to="/auth/register"
            className="link"
          >
            Create New Account
          </Link>
      </form>
    </>
  )
}
