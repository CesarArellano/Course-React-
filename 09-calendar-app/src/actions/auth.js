import { fetchWithoutToken } from "../helpers/fetch";
import { types } from "../types/types";

export const startLogin = ( email, password ) => {
  return async ( dispatch ) => {
    const resp = await fetchWithoutToken('auth', { email, password }, 'POST');
    const decodedData = await resp.json();
    
    if( decodedData.ok ) {
      localStorage.setItem( 'token', decodedData.token );
      localStorage.setItem( 'token-init-date', new Date().getTime() );
      dispatch( login({
        uid: decodedData.uid,
        name: decodedData.name,
      }));
    }
  }
};

const login = ( user ) => ({
  type: types.authLogin,
  payload: user,
});