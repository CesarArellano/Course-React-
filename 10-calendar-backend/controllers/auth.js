const { response } = require('express');
const User = require('../models/User');

const createUser = async ( req, res = response ) => {
  
  const { email, password } = req.body;
  try {
    let user = await User.findOne({ email })

    if( user ) {
      return res.status(400).json({
        ok: false,
        msg: 'El usuario ya existe con ese correo.'
      });
    }

    user = new User( req.body );
    user.save();
    
    res.status(201).json({
      ok: true,
      uid: user.id,
      name: user.name
    });

  } catch(error) {
    console.log(error);
    res.status(500).json({
      ok: false,
      msg: 'Please talk to the admin'
    });
  }
  
}

const userLogin = ( req, res = response ) => {
  const { email, password } = req.body;

  res.json({
    ok: true,
    msg: 'login',
    email,
    password
  });

};

const revalidateToken = ( req, res = response ) => {
  res.json({
    ok: true,
    msg: 'renew'
  });
};

module.exports = {
  createUser,
  userLogin,
  revalidateToken
};