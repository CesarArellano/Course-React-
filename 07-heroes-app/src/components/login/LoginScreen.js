import React, { useContext } from 'react'
import { AuthContext } from '../../auth/AuthContext';
import { types } from '../../types/types';

export const LoginScreen = ({ history }) => {
  const { dispatch } = useContext(AuthContext);
  const handleLogin = () => {
    
    const newLogin = {
      type: types.login,
      payload: {
        name: 'CÃ©sar Arellano'
      }
    };
    
    dispatch(newLogin);
    history.replace('/');
  }

  return (
    <div className="container mt-5">
      <h1>Login</h1>
      <button
        className="btn btn-primary"
        onClick={ handleLogin }
      >
        Sign In
      </button>
    </div>
  )
}
