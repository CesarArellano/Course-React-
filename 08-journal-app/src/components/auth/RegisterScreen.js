import React from 'react'
import { Link } from 'react-router-dom'
import { useDispatch, useSelector } from 'react-redux';
import { useForm } from '../../hooks/useForm'
import validator from 'validator';
import { removeError, setError } from '../../actions/ui';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';

export const RegisterScreen = () => {
  const dispatch = useDispatch();
  const { ui:{ msgError } } = useSelector( state => state); // Recover state data
  
  const [ formValues, handleInputChange ] = useForm({
    name: 'CÃ©sar',
    email: 'cesarmauricio.arellano@gmail.com',
    password: '12345678',
    password2: '12345678'
  });

  const { name, email, password, password2 } = formValues;
  
  const handleRegister = (e) => {
    e.preventDefault();
    
    if( isFormValid() ) {
      dispatch( startRegisterWithEmailPasswordName( email, password, name) )
    }
  };

  const isFormValid = () => {
    if( name.trim().length === 0 ) {
      dispatch(setError('Name is required'));
      return false;
    } else if( !validator.isEmail(email) ){
      dispatch(setError('Email is not valid'));
      return false;
    } else if ( password !== password2 || password.length < 8 ) {
      dispatch(setError('Password should be at least 8 characters and match each other'));
      return false;
    }
    dispatch( removeError() );
    return true;
  };

  return (
    <>
      <div className="center-box">
        <img 
          src="https://image.flaticon.com/icons/png/512/3235/3235042.png"
          alt="Journal App Icon" 
          title="App Icon"
          className="img-icon"
        />
      </div>
      <h3 className="auth__title">Register</h3>
      <form onSubmit={ handleRegister }>
        {
          ( msgError ) && <div className="auth__alert-error animate__animated animate__shakeX">
            { msgError }
          </div>
        }
        
        <input
          type="text"
          placeholder="Name"
          name="name"
          autoComplete="off"
          className="auth__input"
          value={ name }
          onChange={ handleInputChange }
          required
        />
        <input
          type="text"
          placeholder="Email"
          name="email"
          autoComplete="off"
          className="auth__input"
          value={ email }
          onChange={ handleInputChange }
        />

        <input 
          type="password"
          placeholder="Password"
          name="password"
          className="auth__input"
          value={ password }
          onChange={ handleInputChange }
        />

        <input 
          type="password"
          placeholder="Confirm password"
          name="password2"
          className="auth__input"
          value={ password2 }
          onChange={ handleInputChange }
        />

        <div className="center-box">
          <button
            type="submit"
            className="btn btn-primary mb-5"
          >
            Register
          </button>
        </div>
          <Link 
            to="/auth/login"
            className="link"
          >
            Already registered?
          </Link>
      </form>
    </>
  )
}
