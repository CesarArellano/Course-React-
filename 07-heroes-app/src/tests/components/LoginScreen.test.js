import { mount } from 'enzyme';
import { AuthContext } from '../../auth/AuthContext';
import { LoginScreen } from '../../components/login/LoginScreen';
import { types } from '../../types/types';

describe('Testing with <LoginScreen />', () => {
  const contextValue = {
    dispatch: jest.fn()
  }

  const historyMock = {
    push: jest.fn(),
    location: {},
    listen: jest.fn(),
    createHref: jest.fn(),
    replace: jest.fn()
  }

  const wrapper = mount(
    <AuthContext.Provider value={ contextValue }>
      <LoginScreen history={ historyMock } />
    </AuthContext.Provider>
  );

  test('Should show it correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  test('should call dispatch in LoginScreen', () => {
    const handleClick = wrapper.find('button').prop('onClick');
    handleClick();
    expect(contextValue.dispatch).toHaveBeenCalledWith({
      type: types.login,
      payload: {
        name: 'CÃ©sar Arellano'
      }
    });
    expect(historyMock.replace).toHaveBeenCalled();
    localStorage.setItem('lastPath', '/dc');
    handleClick();
    expect(historyMock.replace).toHaveBeenCalledWith('/dc');
  });

});
