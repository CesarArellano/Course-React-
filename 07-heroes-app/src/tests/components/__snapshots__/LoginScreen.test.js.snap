// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing with <LoginScreen /> Should show it correctly 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login
  </h1>
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Sign In
  </button>
</div>
`;
